### PR TITLE
imp interface for SL2RD

### PR DESCRIPTION
**Pull request by xiangformless**
- imp IERC20Payable interface for SL2RD

### CODE REVIEW

**Comment by xiangformless** [comment]
image

**Comment by xiangformless** [comment]
@un7c0rn could you please help to take a look on this test failure? cannot be deployed. The code was successful on my local. Thanks!
image

File: contracts/test/contract.SL2RD.test.js [hunk 1]
```diff
@@ -19,6 +19,28 @@ function calculateSplitIndexUsingPartition(
 }
 
 contract("SL2RD", (accounts) => {
+
+  const mockERC20Address = "0x1234567890abcdef1234567890abcdef12345678"; // Mock ERC20 contract address
+
+  before(async () => {
```
Comment (un7c0rn) [inline contracts/test/contract.SL2RD.test.js:25]: I like these changes, seems much more efficient. We should update the names to distinguish from the shareContract and splitContract names used within existing tests. Perhaps something like this._singletonShareContract or similar.

Comment (un7c0rn) [reply contracts/test/contract.SL2RD.test.js:25]: @xiangformless

File: contracts/test/contract.SL2RD.test.js [hunk 2]
```diff
@@ -963,7 +985,6 @@ contract("SL2RD", (accounts) => {
       const recipientAddress = accounts[NON_OWNER_ADDRESS_INDEX];
       const uniformCollaboratorsIds = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9];
       const operatorAddresses = [accounts[5], accounts[6]];
-      const nonOperatorAddress = accounts[7];
```
Comment (un7c0rn) [inline contracts/test/contract.SL2RD.test.js:966]: @xiangformless Reason for deletion?

**Comment by xiangformless** [comment]
After I reverted all changes on contract.SL2RD.test.js, this PR still failed. Failure details were attached below. @un7c0rn Could you please take a look and let me know if these failures were caused by the implementation of IERC20Payable interface on SL2RD.sol? Shall I make changes on these failed unit tests? or Is there anything missing on SL2RD.sol?
image

File: contracts/test/contract.OperatorRegistry.test.js [hunk 3]
```diff
@@ -113,7 +101,9 @@ contract("OperatorRegistry", (accounts) => {
         verifiedShareOperatorEOAs[i]
       );
 
-      const fundsDelta = newBalance - initialBalance[i];
+      const fundsDelta = web3.utils
```
Comment (un7c0rn) [inline contracts/test/contract.OperatorRegistry.test.js]: @xiangformless In the test results:

```
  Contract: OperatorRegistry                                                                                                                                                                                          
                                                                                                                                                                                                                      
New balance: 100002500000000000000000                                                                                                                                                                                 
Initial balance: 100000000000000000000000                                                                                                                                                                             
Funds delta: 2500000000008782000                                                                                                                                                                                      
    1) Fund the operator addresses                                                                                                                                                                                    
    > No events were emitted                                                                                                                                                                                          
                                                                                                                                                                                                                      
                                                                                                                                                                                                                      
  0 passing (290ms)                                                                                                                                                                                                   
  1 failing                                                                                                                                                                                                           
                                                                                                                                                                                                                      
  1) Contract: OperatorRegistry                                                                                                                                                                                       
       Fund the operator addresses:                                                                                                                                                                                   
     AssertionError: Operator 0 was not correctly funded. 
```

I was able to see that the new balance 100002500000000000000000, and initial balance 100000000000000000000000, were correct. However, the calculated delta was wrong and had the additional value `8782000`. Since these numbers are huge these kinds of errors typically mean there's a problem with the arithmetic implementation, which may show up intermittently. Updating to use BigNumbers fixes the issue.

**Comment by xiangformless** [comment]
@un7c0rn Thanks for fixing the test failure. I pulled the latest code down and rerun the truffle test. I can still see a failure as below. Shall I simply extend the timeout setting in the SL2RD test file? 
image

**Comment by un7c0rn** [comment]
@xiangformless This appears to be passing in the PR.

**Comment by xiangformless** [comment]
@un7c0rn could you please help to approve this PR? So that I can merge it. thanks!

File: contracts/SL2RD.sol [hunk 4]
```diff
@@ -475,4 +482,28 @@ contract SL2RD is
...
+            contractAddress_ != address(0),
+            "Invalid ERC20 contract address"
+        );
+        _erc20ContractAddress = contractAddress_;
+    }
+
+    /// @notice Gets the ERC20 contract address used for payments.
+    function getERC20ContractAddress() external view returns (address) {
+        return _erc20ContractAddress;
+    }
```
Comment (un7c0rn) [inline contracts/SL2RD.sol:499]: nit: linebreak

File: truffle-config.js [hunk 5]
```diff
@@ -91,7 +91,8 @@ module.exports = {
       version: SOLIDITY_COMPILER_VERSION,
       settings: {
         optimizer: {
-          enabled: false,
+          enabled: true, // Default: false
```
Comment (un7c0rn) [inline truffle-config.js]: Are these changes needed?

**Review by un7c0rn — CHANGES_REQUESTED** [review]
Thanks @xiangformless! Minor notes inline.

File: contracts/test/contract.SL2RD.test.js (hunk 1 above)
Comment (xiangformless) [reply contracts/test/contract.SL2RD.test.js:25]: @un7c0rn Should I update the names of this.shareContract, this.operatorRegistry, and this.splitContract to this._singletonShareContract, this._singletonOperatorRegistry, and this._singletonSplitContract for all the places in this SL2RD.test.js file?

File: contracts/test/contract.OperatorRegistry.test.js (hunk 3 above)
Comment (xiangformless) [reply contracts/test/contract.OperatorRegistry.test.js]: @un7c0rn I was trying to make changes based on this comment, but my local truffle test is failing as below 
image
Could you please let me know how to resolve this? I pulled latest code from this branch, and also, this PR is failing again with different reason..thanks

Comment (un7c0rn) [reply contracts/test/contract.OperatorRegistry.test.js]: @xiangformless Does it pass if we restore this file to main?

File: contracts/test/contract.SL2RD.test.js (hunk 1 above)
Comment (un7c0rn) [reply contracts/test/contract.SL2RD.test.js:25]: @xiangformless Yes because the new instance variable `this.shareContract` is an overloaded name that may cause confusion when viewed alongside `shareContract` variables that are not an instance of the singleton object associated with the test.

File: truffle-config.js [hunk 6]
```diff
@@ -16,7 +16,7 @@ module.exports = {
       host: "127.0.0.1",
       port: 8545,
       network_id: "*",
-      gas: 800000000,
+      gas: 80000000000000000,
```
Comment (un7c0rn) [inline truffle-config.js]: @xiangformless Is this required?

Comment (xiangformless) [reply truffle-config.js]: without this change, I got local test failure like below. 
image

File: contracts/test/contract.OperatorRegistry.test.js (hunk 3 above)
Comment (xiangformless) [reply contracts/test/contract.OperatorRegistry.test.js]: after reverting back to main, this file is still getting the same error. as below. 
image
I am curious how you get these tests passed successfully earlier. As I am only making minor changes, I do not think these changes can causes these failures.

File: truffle-config.js (hunk 6 above)
Comment (un7c0rn) [reply truffle-config.js]: @xiangformless SG let's keep.

File: contracts/test/contract.OperatorRegistry.test.js (hunk 3 above)
Comment (un7c0rn) [reply contracts/test/contract.OperatorRegistry.test.js]: @xiangformless I think you removed the optimization changes in this PR when I asked if they were required, perhaps that's what's causing the issue here? The error is that the final binary compilation of the contract is too big for the chain. Optimization should help with this.

Comment (xiangformless) [reply contracts/test/contract.OperatorRegistry.test.js]: @un7c0rn Yes, I added those optimization for debugging purpose. I thought it might be redundant after your comments. Let me put them back to see if it can pass. thanks

Comment (xiangformless) [reply contracts/test/contract.OperatorRegistry.test.js]: @un7c0rn I restored the optimizer as I set before, but the error is not resolved. still same "Number can only safely store up to 53 bits"error,   optimizer: {
          enabled: true, // Default: false
          runs: 1000 // Default: 200
        },

Comment (un7c0rn) [reply contracts/test/contract.OperatorRegistry.test.js]: Ok taking a look @xiangformless!

File: contracts/SL2RD.sol [hunk 7]
```diff
@@ -475,4 +481,29 @@ contract SL2RD is
...
         asset.transferOwnership(msg.sender);
     }
+
+    /// @notice Sets the ERC20 contract address (e.g., for USDC payments).
+    function setERC20ContractAddress(
+        address contractAddress_
+    ) public override afterInit onlyOwner nonReentrant {
+        require(
+            contractAddress_ != address(0),
+            "Invalid ERC20 contract address"
```
Comment (un7c0rn) [inline contracts/SL2RD.sol]: @xiangformless Please follow the convention here of using a `SHARE<number>` error code. See [1] & [2] for example.

[1] https://github.com/formless-eng/share-v1-core/blob/9644cd7b695a8fd06169905123e45607f1ebd4e1/contracts/PFA.sol#L185

[2] https://github.com/formless-eng/share-v1-core/blob/9644cd7b695a8fd06169905123e45607f1ebd4e1/README.md?plain=1#L123

File: contracts/test/contract.OperatorRegistry.test.js (hunk 3 above)
Comment (un7c0rn) [reply contracts/test/contract.OperatorRegistry.test.js]: @xiangformless I've updated the truffle config and am re-kicking the tests now.

File: contracts/SL2RD.sol [hunk 8]
```diff
@@ -475,4 +481,29 @@ contract SL2RD is
...
         asset.transferOwnership(msg.sender);
     }
+
+    /// @notice Sets the ERC20 contract address (e.g., for USDC payments).
+    function setERC20ContractAddress(
+        address contractAddress_
+    ) public override afterInit onlyOwner nonReentrant {
+        require(
+            contractAddress_ != address(0),
+            "SHARE001"
```
Comment (un7c0rn) [inline contracts/SL2RD.sol]: @xiangformless This is the right idea but the error code is wrong. You'll need to make a new one specific to this error.

**Review by un7c0rn — APPROVED** [review]
LGTM!
